sdk/dotnet/Pulumi/Deployment/Deployment_Run.cs

```diff
@@ -87,6 +89,32 @@ public static Task<int> RunAsync(Func<Task<IDictionary<string, object?>>> func)
         public static Task<int> RunAsync<TStack>() where TStack : Stack, new()
             => CreateRunner().RunAsync<TStack>();
 
+        /// <summary>
+        /// Entry point to test a Pulumi application. Deployment will
+        /// instantiate a new stack instance based on the type passed as TStack
+        /// type parameter. This method creates no real resources.
+        /// Note: Currently, unit tests that call <see cref="TestAsync{TStack}"/>
+        /// must run sequentially; parallel execution is not supported.
```

Comment:
Do we have any path to enabling this?  It does seem a likely very significant restriction. 

Also - if this has to be the case for now - should we at least enforce it with a good error message?

Reply:
Not really... the whole .NET SDK implementation relies heavily on static state, which means parallel executions would bite each other.

We had
```
#3399 (comment). Can we get off of using static state. Perhaps have the instance
state computed and passed into the RunAsync method as an appropriate context.
Note: inputs/outputs are difficult here as they look at state like IsPreview
and don't always have a clear way to determine that value.
```
in #3470 but it was never considered a priority.

Enforcing sounds like a good idea.

Reply:
Added the checks around `TestAsync` to throw an exception if multiple parallel executions are detected